### PR TITLE
CDAP-3442 fix deploy failures when a config is given to plain app

### PR DESCRIPTION
**Pull request by albertshau**
Fixes a bug where application deployment would fail if a config
was given to an Application class that extends AbstractApplication
without any special config object.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-RBT432

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java [hunk 1]
```diff
@@ -171,7 +171,11 @@ private String getSpecJson(Application app, final String bundleVersion, final St
         appConfig = (Config) configToken.getRawType().newInstance();
       } else {
         try {
-          appConfig = GSON.fromJson(configString, configToken.getType());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: Can you explain a bit? When is `configToken.getType()` returns if the application has extends without specifying the parameterized type (I guess that's what the fix is about)?

**Comment by gokulavasan** [comment]
@albertshau ``Config`` doesn't contain PluginProperties anymore. Is this issue still present?

**Comment by albertshau** [comment]
it's still an issue without PluginProperties.  What happens is configToken.getType() returns some generic T type so gson deserializes it as a TreeMap.  It then tries to assign that to appConfig and we get a ClassCastException since it's trying to cast TreeMap to Config.

**Comment by chtyim** [comment]
It seems wrong that the configType.getType() returns generic type. How do we create the configToken? The deployment pipeline should be extracting it from the parameterized type and verify it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java [hunk 2]
```diff
@@ -187,4 +184,15 @@ private String getSpecJson(Application app, final String bundleVersion, final St
...
   }
 
+  private Type getConfigType(Class<? extends Application> appClass) {
+    TypeToken<?> configType = TypeToken.of(appClass).resolveType(Application.class.getTypeParameters()[0]);
+    if (Reflections.isResolved(configType.getType())) {
+      return configType.getType();
+    }
+
+    // It has to be Config
+    Preconditions.checkArgument(Config.class == configType.getRawType(), "Application config type not supported. " +
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: Should this check be done here or when we add an app artifact?

Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: thats a good point, this should be checked when deploying an artifact

**Comment by albertshau** [comment]
@gokulavasan @chtyim i've addressed comments, could you take another look?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java [hunk 3]
```diff
@@ -169,13 +167,13 @@ private String getSpecJson(Application app, final String bundleVersion, final St
...
-      TypeToken<?> configToken = typeToken.resolveType(Application.class.getTypeParameters()[0]);
+      Type configType = Artifacts.getConfigType(app.getClass());
       if (Strings.isNullOrEmpty(configString)) {
-        appConfig = (Config) configToken.getRawType().newInstance();
+        //noinspection unchecked
+        appConfig = ((Class<? extends Config>) configType).newInstance();
       } else {
         try {
-          appConfig = GSON.fromJson(configString, configToken.getType());
+          appConfig = GSON.fromJson(configString, Artifacts.getConfigType(app.getClass()));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: Why need to get the config type again? Isn't it the `configType` already?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/Artifacts.java [hunk 4]
```diff
@@ -28,6 +35,25 @@ public static String getFileName(ArtifactId artifactId) {
...
+   */
+  public static Type getConfigType(Class<? extends Application> appClass) {
+    TypeToken<?> configType = TypeToken.of(appClass).resolveType(Application.class.getTypeParameters()[0]);
+    if (Reflections.isResolved(configType.getType())) {
+      return configType.getType();
+    }
+
+    // It has to be Config
+    Preconditions.checkArgument(Config.class == configType.getRawType(), "Application config type not supported. " +
+      "Type must extend Config and cannot be parameterized.");
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/Artifacts.java]: Maybe good to have `configType` as part of the error message.

**Comment by chtyim** [comment]
Just two minor comments. Rest LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java]: yeah this should not be here

**Pull request merged**